File: FindFirstCommonNode.py
#!/usr/bin/python3

# -*- coding:utf-8 -*-
"""

@author: forgotten_liu
@projectName: python_study
@file: FindFirstCommonNode
@time: 2021/1/26 17:49
@IDE: PyCharm
@desc: 输入两个链表，找出它们的第一个公共结点。
（注意因为传入数据是链表，所以错误测试数据的提示是用其他方式显示的，保证传入数据是正确的）
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# @param pHead1 ListNode类
# @param pHead2 ListNode类
# @return ListNode类


# 方法一、分别遍历两个链表，将每个链表存储至对应的列表内，对列表进行 pop 操作，第一对不相同节点的 next 即是公共节点
class Solution:
    def FindFirstCommonNode(self, pHead1, pHead2):
        if not pHead1 and not pHead2:
            return None
        pList1, pList2 = [], []
        tmp = ListNode(None)
        while pHead1:
            pList1.append(pHead1)
            pHead1 = pHead1.next
        while pHead2:
            pList2.append(pHead2)
            pHead2 = pHead2.next
        # 输出 两个列表内数值
        # print([x.val for x in pList1], [x.val for x in pList2])
        for node in range(min(len(pList1), len(pList2))):
            pNode1 = pList1.pop()
            pNode2 = pList2.pop()
            # 输出 弹出值
            # print(pNode1.val, pNode2.val)
            if pNode2.val == pNode1.val:
                tmp = pNode1
            else:
                break
        return tmp


# 方法二、通过 链表 pHead1 与 pHead2 之间互相拼接，使 p1、p2一直循环直到公共节点 停止循环
class Solution1:
    def FindFirstCommonNode(self, pHead1, pHead2):
        if not pHead1 or not pHead2:
            return None
        p1 = pHead1
        p2 = pHead2
        while p1 != p2:
            """
            第一轮循环:
                1, 2, 3, 4, 5, 6, 结束第一轮 p2 = Phead1, p1 = p1.next = 9
            p1: 1, 3, 5, 6, 7, 8   
            p2: 2, 4, 6, 7, 8, 9   
            第二轮循环 p2 = Phead1, p1 = p1.next = 9 对比一次后 p1 = pHead2:
                1, 2(p1=Phead2),   3,  4(p1 = p2 对比结束)
            p1: 9,      2,         4,  6
            p2: 1,      3,         5,  6
            返回 p1 此时 p1 = [6, 7, 8, 9]
            """
            p1 = p1.next if p1 else pHead2
            p2 = p2.next if p2 else pHead1
        print(type(p1))
        return p1


# pHead1 [1, 3, 5, 6, 7, 8, 9]
# pHead2 [2, 4, 6, 7, 8, 9]
pHead1 = ListNode(x=1)
pHead1.next = ListNode(3)
pHead1.next.next = ListNode(5)
pHead1.next.next.next = ListNode(6)
pHead1.next.next.next.next = ListNode(7)
pHead1.next.next.next.next.next = ListNode(8)
pHead1.next.next.next.next.next.next = ListNode(9)

pHead2 = ListNode(2)
pHead2.next = ListNode(4)
pHead2.next.next = ListNode(6)
pHead2.next.next.next = ListNode(7)
pHead2.next.next.next.next = ListNode(8)
pHead2.next.next.next.next.next = ListNode(9)

# s = Solution()
# print(s.FindFirstCommonNode(pHead1, pHead2).val)
s1 = Solution1()
print(s1.FindFirstCommonNode(pHead1, pHead2))

File: FindFirstCommonNode.go
//package main

/*
@author: forgotten_liu
@projectName: Golang_study
@file: FindFirstCommonNode
@time: 2021/1/26 19:05
@IDE: GoLand
@desc: 输入两个链表，找出它们的第一个公共结点。
（注意因为传入数据是链表，所以错误测试数据的提示是用其他方式显示的，保证传入数据是正确的）
*/

package main

import "fmt"

type ListNode struct{
	Val int
	Next *ListNode
}

/**
 *
 * @param pHead1 ListNode类
 * @param pHead2 ListNode类
 * @return ListNode类
 */

// 方法一、存入列表依次 弹出对比
func FindFirstCommonNode( pHead1 *ListNode ,  pHead2 *ListNode ) *ListNode {
	if pHead1 == nil && pHead2 == nil{
		return nil
	}
	var pList2 []ListNode
	var pList1 []ListNode
	for pHead1 != nil{
		pList1 = append(pList1, *pHead1)
		pHead1 = pHead1.Next
	}
	for pHead2 != nil{
		pList2 = append(pList2, *pHead2)
		pHead2 = pHead2.Next
	}
	//for i := 0; i < len(pList1); i ++{
	//	fmt.Println(pList1[i].Val)
	//}
	//for i := 0; i < len(pList2); i ++{
	//	fmt.Println(pList2[i].Val)
	//}
	var tmp *ListNode
	length1, length2 := len(pList1)-1, len(pList2)-1
	for length1 >= 0 && length2 >= 0{
		pNode1, pNode2 := pList1[length1], pList2[length2]
		//fmt.Println(pNode1, pNode2)
		if pNode1.Val == pNode2.Val{
			tmp = &pNode1
		}else{
			break
		}
		length1 -= 1
		length2 -= 1
	}
	return tmp
}

// 方法二、通过 链表 pHead1 与 pHead2 之间互相拼接，使 p1、p2一直循环直到公共节点 停止循环
func FindFirstCommonNode1( pHead1 *ListNode ,  pHead2 *ListNode ) *ListNode {
	if pHead1 == nil || pHead2 == nil{
		return nil
	}
	arr1, arr2 := pHead1, pHead2
	for arr1 != arr2 {
		if arr1 == nil {
			arr1 = pHead2
		}else{
			arr1 = arr1.Next
		}
		if arr2 == nil {
			arr2 = pHead1
		}else {
			arr2 = arr2.Next
		}
	}
	return arr1
}

func main() {
	// pHead1 [1, 3, 5, 6, 7, 8, 9]
	// pHead2 [2, 4, 6, 7, 8, 9]
	pHead1 := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 3,
			Next: &ListNode{
				Val: 5,
				Next: &ListNode{
					Val: 6,
					Next: &ListNode{
						Val: 7,
						Next: &ListNode{
							Val: 8,
							Next: &ListNode{
								Val: 9,
								Next: nil,
							},
						},
					},
				},
			},
		},
	}

	pHead2 := &ListNode{
		Val: 2,
		Next: &ListNode{
			Val: 4,
			Next: &ListNode{
				Val: 6,
				Next: &ListNode{
					Val: 7,
					Next: &ListNode{
						Val: 8,
						Next: &ListNode{
							Val: 9,
							Next: nil,
						},
					},
				},
			},
		},
	}
	pHead3 := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: nil,
				},
			},
		},
	}
	pHead4 := &ListNode{
		Val: 5,
		Next: &ListNode{
			Val: 6,
			Next: &ListNode{
				Val: 7,
				Next: nil,
			},
		},
	}

	arr := FindFirstCommonNode(pHead1, pHead2)
	arr1 := FindFirstCommonNode(pHead3, pHead4)
	fmt.Println(arr, arr1)
	arr2 := FindFirstCommonNode1(pHead1, pHead2)
	arr3 := FindFirstCommonNode1(pHead3, pHead4)
	fmt.Println(arr2, arr3)
}